Add: validation of bundled themes in build workflow

sarsapar1lla opened the pull request:
Added xtask to validate bundled themes in order to detect issues at build time. Basically just loading all the themes are surfacing any warning logs which are emitted.

Main issue with the current implementation in the CI workflow is that it will cause builds on master to fail until the theme validation issues are addressed by their respective maintainers. Alternatively we could add a separate GitHub workflow which only triggers when the `runtime/themes` directory is edited but this felt like overkill in this instance.

Add: xtask for validating bundled themes
Update: build workflow to run theme validation on build

Closes #5709

the-mikedavis reviewed (commented):
I would prefer that we refactor `Theme::load` to return a `Vec` of any warnings (so basically `Result<(Theme, Vec<String>)>`). The current callers of `Theme::load` can then log the warnings and the xtask can print them and error out.

the-mikedavis commented:
For the themes that currently fail to pass this we can comment out the offending keys and link to the issue. They're basically discarded while loading so commenting them out should be equivalent

sarsapar1lla commented:
Updated the implementation to explicitly return validation failures rather than use log captures. This involved adding a new public factory constructor on the `Theme` struct, `Theme::from_keys`, which allows for the creation of a theme from a toml map. This felt less intrusive than altering the return type of `Loader.load` plus allows the `From<Value>` and `Deserialize` impls for `Theme` to still work.

Please let me know if I've misunderstood what you were suggesting!

Regarding the existing themes with validation errors, it looks like some of them are referencing ANSI colours declared in a parent palette which they inherit from. For example, the `gruvbox_dark_hard` theme tries to make use of the "red0" and "bg2" colours declared in the `gruvbox` theme palette but this causes a `Malformed ANSI` error to be returned. Is this expected behaviour for theme inheritance?

sarsapar1lla commented:
I've updated the implementation to return both the theme and any validation failures from the `Loader.load` method and logged the failures at each call site. I've also updated the affected themes and commented out the invalid keys as per your suggestion.

The `theme_check` task now completes successfully and the warnings no longer appear in the logs when selecting themes such as `emacs`.

Let me know what you think @the-mikedavis

the-mikedavis commented on helix-view/src/theme.rs:
```diff
@@ -52,21 +52,24 @@ impl Loader {
 
     /// Loads a theme searching directories in priority order.
-    pub fn load(&self, name: &str) -> Result<Theme> {
+    pub fn load(&self, name: &str) -> Result<(Theme, Vec<String>)> {
```
Since we have many callers that need to handle the warnings let's create a `load_with_warnings` helper that has this function's body and then `load` can call that helper and log the warnings.

sarsapar1lla commented on helix-view/src/theme.rs:
```diff
@@ -53,20 +53,34 @@ impl Loader {
+        let (theme, warnings) = self.load_with_warnings(name)?;
+
+        for warning in warnings {
+            warn!("Theme '{}': {}", name, warning);
```
Logging theme name along with warning

sarsapar1lla commented:
Introduced helper method as suggested. Now that the warnings are being logged in the same place, I updated the warning messages to include the theme name to make it clearer to users which theme is the source of the warning

the-mikedavis reviewed (approved):
Looks good! I'm looking forward to this - I always pull down a theme PR and run it and look at the log and automating this step will be nice
